torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -113,7 +115,80 @@ class GlooStore : public ::gloo::rendezvous::Store {
 
 typedef void (*ReduceFunc)(void*, const void*, const void*, size_t);
 
-template <typename T>
+template <
+    typename T,
+    typename std::enable_if<!std::is_integral<T>::value, int>::type = 0>
+ReduceFunc toFunction(const ReduceOp& r) {
+  switch (r) {
+    case ReduceOp::SUM:
+      return ReduceFunc(&::gloo::sum<T>);
+    case ReduceOp::PRODUCT:
+      return ReduceFunc(&::gloo::product<T>);
+    case ReduceOp::MIN:
+      return ReduceFunc(&::gloo::min<T>);
+    case ReduceOp::MAX:
+      return ReduceFunc(&::gloo::max<T>);
+    case ReduceOp::BAND:
+      throw std::runtime_error(
+          "Cannot use ReduceOp.BAND with non-integral dtype");
+      break;
+    case ReduceOp::BOR:
+      throw std::runtime_error(
+          "Cannot use ReduceOp.BOR with non-integral dtype");
+      break;
+    case ReduceOp::BXOR:
+      throw std::runtime_error(
+          "Cannot use ReduceOp.BXOR with non-integral dtype");
+      break;
+    case ReduceOp::UNUSED:
+      break;
+  }
+
+  throw std::runtime_error("Unhandled ReduceOp");
+}
+
+// Bitwise AND with SFINAE guard for integral types.
+template <
+    typename T,
+    typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+void band(void* c, const void* a, const void* b, size_t n) {
+  auto tc = static_cast<T*>(c);
+  auto ta = static_cast<const T*>(a);
+  auto tb = static_cast<const T*>(b);
+  for (size_t i = 0; i < n; i++) {
```

Comment:
Do we need to validate that we won't get indexing errors here, or does Gloo's reduce function provide us with the correct `n`?